library/src/main/java/com/google/android/exoplayer2/text/eia608/Eia608Decoder.java

```diff
@@ -414,68 +453,102 @@ private void handleMiscCode(byte cc2) {
 
     switch (cc2) {
       case CTRL_ERASE_DISPLAYED_MEMORY:
-        captionString = null;
         if (captionMode == CC_MODE_ROLL_UP || captionMode == CC_MODE_PAINT_ON) {
-          captionStringBuilder.setLength(0);
+          currentCue = new Eia608CueBuilder();
+          cues.clear();
         }
+        subtitle.setCues(Collections.<Cue>emptyList());
         return;
       case CTRL_ERASE_NON_DISPLAYED_MEMORY:
-        captionStringBuilder.setLength(0);
+        currentCue = new Eia608CueBuilder();
+        cues.clear();
         return;
       case CTRL_END_OF_CAPTION:
-        captionString = getDisplayCaption();
-        captionStringBuilder.setLength(0);
+        renderCues();
+        cues.clear();
         return;
       case CTRL_CARRIAGE_RETURN:
-        maybeAppendNewline();
+        // Each time a Carriage Return is received, the text in the top row of the window is erased
+        // from memory and from the display. The remaining rows of text are each rolled up into the
+        // next highest row in the window, leaving the base row blank and ready to accept new text.
+        if (captionMode == CC_MODE_ROLL_UP) {
+          for (Eia608CueBuilder cue : cues) {
+            // Roll up all the other rows.
+            if (!cue.rollUp()) {
+              cues.remove(cue);
```

Comment:
This code can cause `ConcurrentModificationException` when `rollUp()` returns false.